lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -253,94 +251,95 @@ def log.trace?
-        metadata = {
-          'docker' => {
-            'container_id' => container_id
-          },
-          'kubernetes' => get_metadata_for_record(match_data, container_id, create_time_from_record(es.first[1]), batch_miss_cache)
-        }
+    def create_time_from_record(record, internal_time)
+      time_key = @time_fields.detect{ |ii| record.has_key?(ii) }
+      time = record[time_key]
+      if time.nil? || time.chop.empty?
+        return internal_time
       end
-
-      es.each do |time, record|
-        record = record.merge(Marshal.load(Marshal.dump(metadata))) if metadata
-        new_es.add(time, record)
+      if time_key == '_SOURCE_REALTIME_TIMESTAMP' || time_key == '__REALTIME_TIMESTAMP'
+        timei= time.to_i
+        return Time.at(timei / 1000000, timei % 1000000)
       end
-      dump_stats
-      new_es
+      return Time.parse(time)
     end
 
-    def filter_stream_from_journal(tag, es)
+    def filter_stream(tag, es)
       return es if (es.respond_to?(:empty?) && es.empty?) || !es.is_a?(Fluent::EventStream)
-      new_es = MultiEventStream.new
+      new_es = Fluent::MultiEventStream.new
+      tag_match_data = tag.match(@tag_to_kubernetes_name_regexp_compiled) unless @use_journal
+      tag_metadata = nil
       batch_miss_cache = {}
       es.each do |time, record|
-        metadata = nil
-        if record.has_key?('CONTAINER_NAME') && record.has_key?('CONTAINER_ID_FULL')
-          metadata = record['CONTAINER_NAME'].match(@container_name_to_kubernetes_regexp_compiled) do |match_data|
-           container_id = record['CONTAINER_ID_FULL']
-            metadata = {
-              'docker' => {
-                'container_id' => container_id
-              },
-              'kubernetes' => get_metadata_for_record(match_data, container_id, create_time_from_record(record), batch_miss_cache)
-            }
-
-            metadata
-          end
-          unless metadata
-            log.debug "Error: could not match CONTAINER_NAME from record #{record}"
-            @stats.bump(:container_name_match_failed)
-          end
-        elsif record.has_key?('CONTAINER_NAME') && record['CONTAINER_NAME'].start_with?('k8s_')
-          log.debug "Error: no container name and id in record #{record}"
-          @stats.bump(:container_name_id_missing)
+        if tag_match_data && tag_metadata.nil?
+          tag_metadata = get_metadata_for_record(tag_match_data['namespace'], tag_match_data['pod_name'], tag_match_data['container_name'],
+            tag_match_data['docker_id'], create_time_from_record(record, time), batch_miss_cache)
+        end
+        metadata = Marshal.load(Marshal.dump(tag_metadata)) if tag_metadata
+        if (@use_journal || @use_journal.nil?) &&
+          (j_metadata = get_metadata_for_journal_record(record, time, batch_miss_cache))
+          metadata = j_metadata
+        end
+        if @lookup_from_k8s_field && record.has_key?('kubernetes') && record.has_key?('docker') &&
```

Comment:
I don't understand why the record would ever have any of the information populated.   We use the various regex to pull the basic info from either the file path or journal and add them to  the record.  I would expect we still need to do that here.  Did we change something such that records now already include namespace, pod name, container id, and container name?  Was there a change to modify a parser?

Reply:
The point is that this plugin should only know how to look up pod and namespace metadata given a pod name, namespace name, container name.  Why should this plugin have deep knowledge of the container log file format or the CONTAINER_NAME format?  That information could be extracted elsewhere, and passed into this filter plugin via `kubernetes.namespace_name`, etc.

Reply:
Agree. but thats now how it works.  Does it come in via a parser? ... and its metadata... the most basic metadata available which is the point of the plugin.  is it unreasonable to suggest it figures it out.  or...  again... is that the 'parser' that does that?

Reply:
The data could come in via a parser - you could use a record_transformer to parse out the fields from the tag or CONTAINER_NAME - in fact we already do something like this in the mux case, where we parse the incoming tag, then convert it to CONTAINER_NAME for the sole purpose of having the metadata in the right format for the k8s filter plugin - https://github.com/openshift/origin-aggregated-logging/blob/master/fluentd/configs.d/input-post-forward-mux.conf#L30

Reply:
So we are writing this code in anticipation of some 'upstream' bit in the pipeline being able to provide that information?

Reply:
yes - I would be able to take advantage of it immediately in mux